authority/provisioner/azure.go

```diff
@@ -0,0 +1,299 @@
+	if err != nil {
+		return "", errors.Wrap(err, "error getting identity token, are you in a Azure VM?")
+	}
+	defer resp.Body.Close()
+
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", errors.Wrap(err, "error reading identity token response")
+	}
+	if resp.StatusCode >= 400 {
+		return "", errors.Errorf("error getting identity token: status=%d, response=%s", resp.StatusCode, b)
+	}
+
+	var identityToken azureIdentityToken
+	if err := json.Unmarshal(b, &identityToken); err != nil {
+		return "", errors.Wrap(err, "error unmarshaling identity token response")
+	}
+
+	return identityToken.AccessToken, nil
+}
+
+// Init validates and initializes the Azure provisioner.
+func (p *Azure) Init(config Config) (err error) {
+	switch {
+	case p.Type == "":
+		return errors.New("provisioner type cannot be empty")
+	case p.Name == "":
+		return errors.New("provisioner name cannot be empty")
+	case p.TenantID == "":
+		return errors.New("provisioner tenantId cannot be empty")
+	case p.Audience == "": // use default audience
+		p.Audience = azureDefaultAudience
+	}
+	// Initialize config
+	p.assertConfig()
+
+	// Update claims with global ones
+	if p.claimer, err = NewClaimer(p.Claims, config.Claims); err != nil {
+		return err
+	}
+
+	// Decode and validate openid-configuration endpoint
+	if err := getAndDecode(p.config.oidcDiscoveryURL, &p.oidcConfig); err != nil {
+		return err
+	}
+	if err := p.oidcConfig.Validate(); err != nil {
+		return errors.Wrapf(err, "error parsing %s", p.config.oidcDiscoveryURL)
+	}
+	// Get JWK key set
+	if p.keyStore, err = newKeyStore(p.oidcConfig.JWKSetURI); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// AuthorizeSign validates the given token and returns the sign options that
+// will be used on certificate creation.
+func (p *Azure) AuthorizeSign(token string) ([]SignOption, error) {
+	jwt, err := jose.ParseSigned(token)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error parsing token")
+	}
+	if len(jwt.Headers) == 0 {
+		return nil, errors.New("error parsing token: header is missing")
+	}
+
+	var found bool
+	var claims azurePayload
+	keys := p.keyStore.Get(jwt.Headers[0].KeyID)
+	for _, key := range keys {
+		if err := jwt.Claims(key.Public(), &claims); err == nil {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return nil, errors.New("cannot validate token")
+	}
+
+	if err := claims.ValidateWithLeeway(jose.Expected{
+		Audience: []string{p.Audience},
+		Issuer:   p.oidcConfig.Issuer,
+		Time:     time.Now(),
+	}, 1*time.Minute); err != nil {
+		return nil, errors.Wrap(err, "failed to validate payload")
+	}
+
+	// Validate TenantID
+	if claims.TenantID != p.TenantID {
+		return nil, errors.New("validation failed: invalid tenant id claim (tid)")
+	}
+
+	re := azureXMSMirIDRegExp.FindStringSubmatch(claims.XMSMirID)
+	if len(re) == 0 {
+		return nil, errors.Errorf("error parsing xms_mirid claim: %s", claims.XMSMirID)
+	}
+	group, name := re[2], re[3]
```

Comment:
making assumption about the length of `re` here. Maybe check if it is at least 3 in the above length check.

Reply:
If the RE succeeds it will always contain the same number of elements, if not it will contain 0.

Reply:
ok. I feel like checking that it has the exact # of elements you expect would future proof it, but up to you.

Reply:
sure I'll change it